### PR TITLE
Finished the test problems.

### PR DESCRIPTION
**Pull request by MeArio**
### Running:

Run the tests in a terminal with:
`dotnet test`

### Dependencies
Project 'ProblemsSolution' has the following package references
   [netstandard2.0]:

| Top-level Package      | Resolved |
| --------------------------- | ------------ |
| NETStandard.Library | 2.0.3        |

Project 'ProblemsSolution.Tests' has the following package references
   [netcoreapp3.0]:

| Top-level Package         | Resolved |
| ------------------------------| ------------ |
| Microsoft.NET.Test.Sdk | 16.2.0      |
| nunit                              | 3.12.0     |
| NUnit3TestAdapter        | 3.13.0     |

### CODE REVIEW

File: ProblemsSolution/ProblemsSolution.cs [hunk 1]
```diff
@@ -0,0 +1,60 @@
+﻿using System;
+
+namespace ProblemsSolution
+{
+    public class ArrayProblemsSolution
+    {
+        public static int SumArrayEvenNumbers(int[] numbers)
+        {
+            int sum = 0;
+            foreach(int number in numbers)
```
Comment (alex-damian-negru) [inline ProblemsSolution/ProblemsSolution.cs]: This might be simplified. Ever heard of [map](https://blog.submain.com/csharp-functional-programming/)?

File: ProblemsSolution/ProblemsSolution.cs [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+            int arraySum = 0;
+
+            for(int i=1; i <= numbers.Length; i++){
+                indexSum += i;
+                arraySum += numbers[i - 1];
+            }
+
+            if(indexSum == arraySum)
+            {
+                return false;
```
Comment (alex-damian-negru) [inline ProblemsSolution/ProblemsSolution.cs]: The ternary operator could cut up a few lines of code here.

File: ProblemsSolution.Tests/ProblemsSolutionTest.cs [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+namespace ProblemsSolution.Tests
+{
+    [TestFixture]
+    public class Tests
+    {
+
+        [TestCase (new int[] {1,2,3,4,5}, ExpectedResult=6)]
+        [TestCase (new int[] {12, 22, 1, 3}, ExpectedResult=34)]
+        [TestCase (new int[] {3, 7, 0, 1}, ExpectedResult=0)]
+        public int SumArrayEvenNumbersTest(int[] numbers)
```
Comment (alex-damian-negru) [inline ProblemsSolution.Tests/ProblemsSolutionTest.cs:11]: Is it possible to add a short description of each test case? One must be able to see quickly what each test is about, so analyzing every input and extract the description from there is not always great.

File: ProblemsSolution/ProblemsSolution.cs [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+            {
+                if(number % 2 == 0) sum += number;
+            }
+
+            return sum;
+        }
+
+        public static int[] BigestTwoNumbersArray(int[] numbers)
+        {
+            if(numbers.Length == 0) return new int[] {};
```
Comment (alex-damian-negru) [inline ProblemsSolution/ProblemsSolution.cs:16]: Why do you need this guard in place?

**Review by alex-damian-negru — COMMENTED** [review]
Check this gem out regarding commit messages: https://chris.beams.io/posts/git-commit/

File: ProblemsSolution.Tests/ProblemsSolutionTest.cs [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+        [TestCase (new int[] {3}, ExpectedResult=new int[] {3})]
+        [TestCase (new int[] {7, 3}, ExpectedResult=new int[] {3, 7})]
+        [TestCase (new int[] {}, ExpectedResult=new int[] {})]
+        public int[] BigestTwoNumbersArrayTest(int[] numbers)
+        {
+            return ProblemsSolution.ArrayProblemsSolution.BigestTwoNumbersArray(numbers);
+        }
+        [TestCase (new int[] {1, 2, 3, 4, 5}, ExpectedResult=false)]
+        [TestCase (new int[] {1, 2, 3, 2, 5, 6}, ExpectedResult=true)]
+        [TestCase (new int[] {2, 3, 3, 7, 8, 5, 4, 7}, ExpectedResult=true)]
```
Comment (adrianbunea) [inline ProblemsSolution.Tests/ProblemsSolutionTest.cs]: Add this test case and run the tests again [TestCase (new int[] {1, 1, 3, 5, 5}, ExpectedResult=true)]

**Review by alexandru-calinoiu — COMMENTED** [review]
Please update the pr description to provide us with minimum instructions on how to run the code and it's dependencies versions.

File: ProblemsSolution/ProblemsSolution.cs (hunk 1 above)
Comment (MeArio) [reply ProblemsSolution/ProblemsSolution.cs]: I wasn't sure if I was allowed to use LINQ or not so I decided to do it the tedious way. Also not sure how I would do this with map? I'm thinking about filtering even numbers and then reducing with a sum function.

**Comment by MeArio** [comment]
> Please update the pr description to provide us with minimum instructions on how to run the code and it's dependencies versions.

Done

File: ProblemsSolution/ProblemsSolution.cs (hunk 1 above)
Comment (MeArio) [reply ProblemsSolution/ProblemsSolution.cs]: > This might be simplified. Ever heard of [map](https://blog.submain.com/csharp-functional-programming/)?

I've simplified it here aefbbf8c1c711595bcbe35054bb1b342993a8d8c

File: ProblemsSolution.Tests/ProblemsSolutionTest.cs (hunk 3 above)
Comment (MeArio) [reply ProblemsSolution.Tests/ProblemsSolutionTest.cs:11]: > Is it possible to add a short description of each test case? One must be able to see quickly what each test is about, so analyzing every input and extract the description from there is not always great.

Done 80f307ae3cfcdd0b3f05df8f44780a5638854953.

File: ProblemsSolution/ProblemsSolution.cs (hunk 4 above)
Comment (MeArio) [reply ProblemsSolution/ProblemsSolution.cs:16]: > Why do you need this guard in place?

Because otherwise the method would return an array that contains 2 elements with value of Int32.MinValue.

File: ProblemsSolution/ProblemsSolution.cs (hunk 2 above)
Comment (MeArio) [reply ProblemsSolution/ProblemsSolution.cs]: > The ternary operator could cut up a few lines of code here.

Fixed fb48401177edc8163c9370bff510a8dc81099111.

File: ProblemsSolution.Tests/ProblemsSolutionTest.cs (hunk 5 above)
Comment (MeArio) [reply ProblemsSolution.Tests/ProblemsSolutionTest.cs]: > Add this test case and run the tests again [TestCase (new int[] {1, 1, 3, 5, 5}, ExpectedResult=true)]

Yeah, my solution doesn't handle that edge case I will have to think of a better one.

File: ProblemsSolution/ProblemsSolution.cs [hunk 6]
```diff
@@ -0,0 +1,50 @@
+﻿using System;
+using System.Linq;
+
+namespace ProblemsSolution
+{
+    public class ArrayProblemsSolution
+    {
+        public static int SumArrayEvenNumbers(int[] numbers)
+        {
+            int sum = numbers.Where(number => number % 2 == 0).Aggregate((a, b) => a+b);
```
Comment (alexandru-calinoiu) [inline ProblemsSolution/ProblemsSolution.cs]: This can be done nicer,

Can you levreag extension (or whatever mecanism c# has) to create an .even method 
Also you should be able to simply use `.Sum`

File: ProblemsSolution/ProblemsSolution.cs [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+
+        public static bool DuplicatesInArray(int[] numbers)
+        {
+            int indexSum = 0;
+            int arraySum = 0;
+
+            for(int i=1; i <= numbers.Length; i++){
+                indexSum += i;
+                arraySum += numbers[i - 1];
+            }
```
Comment (alexandru-calinoiu) [inline ProblemsSolution/ProblemsSolution.cs]: Not the right solution

It will fail for a simple array like: [2, 2, 2]

File: ProblemsSolution.Tests/ProblemsSolutionTest.cs [hunk 8]
```diff
@@ -0,0 +1,32 @@
+using NUnit.Framework;
+
+namespace ProblemsSolution.Tests
+{
+    [TestFixture]
+    public class Tests
+    {
+
+        [TestCase (new int[] {1,2,3,4,5}, ExpectedResult=6, Description="Check for array with even numbers.")]
+        [TestCase (new int[] {3, 7, 0, 1}, ExpectedResult=0, Description="Check for array with even numbers and a zero.")]
```
Comment (alexandru-calinoiu) [inline ProblemsSolution.Tests/ProblemsSolutionTest.cs:10]: The description is incorrect.

File: ProblemsSolution.Tests/ProblemsSolutionTest.cs [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+        public int SumArrayEvenNumbersTest(int[] numbers)
+        {
+            return ProblemsSolution.ArrayProblemsSolution.SumArrayEvenNumbers(numbers);
+        }
+
+        [TestCase (new int[] {12, 22, 1, 3}, ExpectedResult=new int[] {12, 22}, Description="Check for array with more than 2 elements.")]
+        [TestCase (new int[] {3}, ExpectedResult=new int[] {3}, Description="Check for single element array.")]
+        [TestCase (new int[] {7, 3}, ExpectedResult=new int[] {3, 7}, Description="Check for 2 element array.")]
+        [TestCase (new int[] {}, ExpectedResult=new int[] {}, Description="Check for empty array.")]
```
Comment (alexandru-calinoiu) [inline ProblemsSolution.Tests/ProblemsSolutionTest.cs:19]: How about test cases with negative numbers?

File: ProblemsSolution.Tests/ProblemsSolutionTest.cs [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+        public int[] BigestTwoNumbersArrayTest(int[] numbers)
+        {
+            return ProblemsSolution.ArrayProblemsSolution.BigestTwoNumbersArray(numbers);
+        }
```
Comment (alexandru-calinoiu) [inline ProblemsSolution.Tests/ProblemsSolutionTest.cs:25]: Style looks consistent except this missing empty new line, Wonder what happened here?

**Review by alexandru-calinoiu — COMMENTED** [review]
A decent try, please see my comments.

Looks like for the first solution you skiped over the TDD part.

Looking further down over the commits, this is not really TDD, take a look at this guy to see what I was really expecting https://www.youtube.com/watch?v=qkblc5WRn-U

Some more comments on your commits, they don't seem to be consistent
I almost like: Add descriptions to each test case and remove 1 redundant test 
Don't really like (because is not consistent): Added solution for Simple Problem #3
A good readup about commit messages https://chris.beams.io/posts/git-commit/

File: ProblemsSolution.Tests/ProblemsSolutionTest.cs (hunk 10 above)
Comment (MeArio) [reply ProblemsSolution.Tests/ProblemsSolutionTest.cs:25]: I didn't notice it.

**Comment by MeArio** [comment]
> A decent try, please see my comments.
> 
> Looks like for the first solution you skiped over the TDD part.
> 
> Looking further down over the commits, this is not really TDD, take a look at this guy to see what I was really expecting https://www.youtube.com/watch?v=qkblc5WRn-U
> 
> Some more comments on your commits, they don't seem to be consistent
> I almost like: Add descriptions to each test case and remove 1 redundant test
> Don't really like (because is not consistent): Added solution for Simple Problem #3
> A good readup about commit messages https://chris.beams.io/posts/git-commit/

> A decent try, please see my comments.
> 
> Looks like for the first solution you skiped over the TDD part.
> 
> Looking further down over the commits, this is not really TDD, take a look at this guy to see what I was really expecting https://www.youtube.com/watch?v=qkblc5WRn-U
> 
> Some more comments on your commits, they don't seem to be consistent
> I almost like: Add descriptions to each test case and remove 1 redundant test
> Don't really like (because is not consistent): Added solution for Simple Problem #3
> A good readup about commit messages https://chris.beams.io/posts/git-commit/

I will give that video a watch as soon as I have time.

I should have probably separated that commit in 2 different ones.

Yeah I know the solution is useless I've never been good at these type of problems that's why I never got into competitive programming.

Someone else suggested that read on this PR that's why I've been changing my commit message style and why the commit history is inconsistent.

File: ProblemsSolution/ProblemsSolution.cs (hunk 6 above)
Comment (MeArio) [reply ProblemsSolution/ProblemsSolution.cs]: Done a50351dd8b44e6452baf759fd43fda28b05561e0.

File: ProblemsSolution/obj/ProblemsSolution.csproj.nuget.cache [hunk 11]
```diff
@@ -0,0 +1,5 @@
+{
```
Comment (alexandru-calinoiu) [inline ProblemsSolution/obj/ProblemsSolution.csproj.nuget.cache]: I am preatty sure this `obj` folder should have been excluded.

Comment (MeArio) [reply ProblemsSolution/obj/ProblemsSolution.csproj.nuget.cache]: Fixed I probably forgot to remove the cached files after fixing the .gitignore.

File: ProblemsSolution.Tests/ProblemsSolutionTest.cs (hunk 10 above)
Comment (MeArio) [reply ProblemsSolution.Tests/ProblemsSolutionTest.cs:25]: Fixed af59a0081a4531bc447a1c0ef6d2af8b1b5da9f9.

File: ProblemsSolution/ProblemsSolution.cs (hunk 7 above)
Comment (MeArio) [reply ProblemsSolution/ProblemsSolution.cs]: Fixed: 3d9a558d4d79936dcaf2cd33653540b7dbe1d432 not sure if I should have made a Product extension method since it didn't feel needed.

File: ProblemsSolution.Tests/ProblemsSolutionTest.cs (hunk 9 above)
Comment (MeArio) [reply ProblemsSolution.Tests/ProblemsSolutionTest.cs:19]: > How about test cases with negative numbers?

Yeah it seems that when it comes to negative numbers the array order is reversed, will fix.

Comment (MeArio) [reply ProblemsSolution.Tests/ProblemsSolutionTest.cs:19]: Fixed 45ff35598dc7149787063f58f7e9d84a406b16b2.

File: ProblemsSolution.Tests/ProblemsSolutionTest.cs (hunk 8 above)
Comment (MeArio) [reply ProblemsSolution.Tests/ProblemsSolutionTest.cs:10]: Fixed 3a3cecea212765927f6183cdb4f8273d4c578ee9.

**Comment by alexandru-calinoiu** [comment]
Thank you, will followup via email.
